HIRS_ProvisionerTPM2/src/Process.cpp

```diff
@@ -131,3 +141,62 @@ string Process::run(const string& executable,
     }
     return str;
 }
+
+/**
+ * Static function to check if a specified process is currently running in the
+ * local environment.
+ *
+ * @param executable the executable to check is running
+ * @return true, if executable is found to be running / false, otherwise
+ */
+bool Process::isRunning(const string& executable) {
```

Comment:
I like the `pgrep` suggestion and wouldn't bother using the others, but `pgrep` has its own problem in that it returns a non-zero error code when no processes are matched. Since we'd likely be using the `RUN_PROCESS_OR_THROW` macro, that would result in an Exception being thrown, which requires a `try/catch` for what could easily be handled by an `if/then` statement.

Additionally, upon checking the `man` page of the `pgrep` utility, it would appear that they are checking the `/proc/_pid_/stat` file unless you provide the `-f` flag that shifts it to use the `/proc/_pid_/cmdline` file. This is exactly what this function currently does without relying on yet another external utility. Rather than refactoring the `Process` class further beyond the scope of this ticket to prevent an Exception from being thrown for expected behavior. This seems to be a good enough solution than spend more time on it.

Additionally, a number of tests have been written testing this method. If there are any suggestions regarding further tests, I'm more than happy to write up additional test cases.

Reply:
1. You don't have to use RUN_PROCESS_OR_THROW, especially since it does not make sense in this case.
2. I'd prefer to use tools that have some pedigree rather than trying to rewrite them because edge cases have been dealt with.
3. What is the advantage of searching `/proc/_pid_/cmdline` over `/proc/_pid_/stat`?

Reply:
1. The only option I have since `isRunning` is a `static` method is `run`, which is just what `RUN_PROCESS_OR_THROW` wraps. Changing `isRunning` to be a non-static method seems like a poor design decision since you shouldn't need a `Process` object to determine if a system process is running when `Process` is a utility class of sorts. `isRunning` should belong to the class.
2. Typically I would agree, but in this case, this isn't an overtly complex scenario. The only truly odd edge case is if you're examining a specific process that terminates when you're about to check its `stat` or `cmdline` file. This would of course delete the file that you thought was there. For that case, there's a try/catch as the backup to ensure the program doesn't enter some strange realm due to this potential race condition. Instead it just continues onto the next process in the `/proc` file.
3. Per the `pgrep` utility's `man` page, the `cmdline` file is useful for situations where the process name may be longer than 15 characters. On some systems, this is a truncation point for the `stat` file, but the `cmdline` file will have the complete command that was run to execute the process. Due to the difference in content between the two files, `stat` is preferred for most cases as it has little to no chance of providing a false positive, whereas the `cmdline` increases the chance of a false positive, but is a necessary alternative failing `stat`.

Reply:
Per the other discussion, this approach does not work when support for `cmdline` is desired.